drivers/flash/flash_realtek_rts5912.c

```diff
@@ -0,0 +1,768 @@
+	} instruction;
+	struct {
+		SPIC_BUS_WIDTH_t bus_width; /* Bus width for the address */
+		SPIC_ADDRESS_SIZE_t size;   /* Address size */
+		uint32_t value;             /* Address value */
+		uint8_t disabled;           /* Address phase skipped if disabled is set to true */
+	} address;
+	struct {
+		SPIC_BUS_WIDTH_t bus_width; /* Bus width for alternative */
+		uint8_t size;               /* Alternative size */
+		uint32_t value;             /* Alternative value */
+		uint8_t disabled; /* Alternative phase skipped if disabled is set to true */
+	} alt;
+	uint8_t dummy_count; /* Dummy cycles count */
+	struct {
+		SPIC_BUS_WIDTH_t bus_width; /* Bus width for data */
+	} data;
+};
+
+struct flash_rts5912_dev_data {
+	struct k_sem sem;
+	struct qspi_cmd command_default;
+};
+
+static const uint8_t user_addr_len[] = {
+	[SPIC_CFG_ADDR_SIZE_8] = 1,
+	[SPIC_CFG_ADDR_SIZE_16] = 2,
+	[SPIC_CFG_ADDR_SIZE_24] = 3,
+	[SPIC_CFG_ADDR_SIZE_32] = 4,
+};
+
+static const struct flash_parameters flash_rts5912_parameters = {
+	.write_block_size = FLASH_WRITE_BLK_SZ,
+	.erase_value = 0xff,
+};
+
+static int config_command(struct qspi_cmd *command, uint8_t cmd, uint32_t addr,
+			  SPIC_ADDRESS_SIZE_t addr_size, uint8_t dummy_count)
+{
+	int ret = 0;
+
+	switch (cmd) {
+	case SPI_NOR_CMD_WREN:
+	case SPI_NOR_CMD_WRDI:
+	case SPI_NOR_CMD_WRSR:
+	case SPI_NOR_CMD_RDID:
+	case SPI_NOR_CMD_RDSR:
+	case SPI_NOR_CMD_RDSR2:
+	case SPI_NOR_CMD_CE:
+	case SPI_NOR_CMD_4BA:
+	case FLASH_CMD_EX4B:
+	case FLASH_CMD_EXTNADDR_WREAR:
+	case FLASH_CMD_EXTNADDR_RDEAR:
+	case SPI_NOR_CMD_RESET_EN:
+	case SPI_NOR_CMD_RESET_MEM:
+		command->address.disabled = 1;
+		command->data.bus_width = SPIC_CFG_BUS_SINGLE;
+		break;
+	case SPI_NOR_CMD_READ:
+	case SPI_NOR_CMD_READ_FAST:
+	case SPI_NOR_CMD_SE:
+	case SPI_NOR_CMD_BE:
+	case FLASH_CMD_RDSFDP:
+	case SPI_NOR_CMD_PP:
+		command->address.disabled = 0;
+		command->address.bus_width = SPIC_CFG_BUS_SINGLE;
+		command->data.bus_width = SPIC_CFG_BUS_SINGLE;
+		break;
+	case SPI_NOR_CMD_DREAD:
+		command->address.disabled = 0;
+		command->address.bus_width = SPIC_CFG_BUS_SINGLE;
+		command->data.bus_width = SPIC_CFG_BUS_DUAL;
+		break;
+	case SPI_NOR_CMD_QREAD:
+		command->address.disabled = 0;
+		command->address.bus_width = SPIC_CFG_BUS_SINGLE;
+		command->data.bus_width = SPIC_CFG_BUS_QUAD;
+		break;
+	case SPI_NOR_CMD_2READ:
+		command->address.disabled = 0;
+		command->address.bus_width = SPIC_CFG_BUS_DUAL;
+		command->data.bus_width = SPIC_CFG_BUS_DUAL;
+		break;
+	case SPI_NOR_CMD_4READ:
+	case SPI_NOR_CMD_PP_1_4_4:
+		command->address.disabled = 0;
+		command->address.bus_width = SPIC_CFG_BUS_QUAD;
+		command->data.bus_width = SPIC_CFG_BUS_QUAD;
+		break;
+	default:
+		ret = -EINVAL;
+		break;
+	}
+
+	command->instruction.value = cmd;
+	command->address.size = addr_size;
+	command->address.value = addr;
+	command->dummy_count = dummy_count;
+
+	return ret;
+}
+
+static int spic_wait_finish(void)
```

Comment:
Pass the `const struct device *dev` to all these local routines that need to access the registers.
 ```suggestion
static int spic_wait_finish(const struct device *dev) {
    const struct flash_rts5912_dev_config *config = dev->config;
    struct spi_type *spic_reg = config->spic_reg;
}
```